File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int n, m;
    void four_dirn(vector<vector<char>>& grid, int i, int j){
        if (i>=0 && i<n && j>=0 && j<m && grid[i][j] == '1'){
            grid[i][j] = '2';
            four_dirn(grid, i, j+1);
            four_dirn(grid, i+1, j);
            four_dirn(grid, i, j-1);
            four_dirn(grid, i-1, j);
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        n = grid.size();
        m = grid[0].size();
        int island_count = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if (grid[i][j] == '1'){
                    island_count++;
                    four_dirn(grid, i, j);
                }
                
            }
        }
        return island_count;
    }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])

        def bfs(i, j):
            if 0 <= i < n and 0 <= j < m and grid[i][j] == "1":
                grid[i][j] = "2"

                bfs(i+1, j)
                bfs(i-1, j)
                bfs(i, j+1)
                bfs(i, j-1)
        
        no_of_islands = 0

        for i in range(n):
            for j in range(m):
                if grid[i][j] == "1":
                    no_of_islands += 1
                    bfs(i, j)
        return no_of_islands